models/yang/sonic/sonic-acl.yang

```diff
@@ -0,0 +1,233 @@
+
+				leaf aclname {
+					type leafref {
+						path "../../../ACL_TABLE/ACL_TABLE_LIST/aclname";
+					}
+				} 
+
+				leaf rulename {
+					type string {
+						pattern '[a-zA-Z0-9]{1}([-a-zA-Z0-9_]{1,63})' {
+							error-app-tag rulename-invalid;
+						}
+					}
+				}
+
+				leaf PRIORITY {
+					type uint16 {
+						range "1..65535"{
+							error-app-tag priority-invalid-range;
+							error-message "Invalid ACL rule priority.";
+						}
+					}
+				}
+
+				leaf RULE_DESCRIPTION {
+					type string {
+						length 1..255 {
+							error-app-tag ruledesc-invalid-length;
+						}
+					}
+				}
+
+				leaf PACKET_ACTION {
+					mandatory true;
+					type enumeration {
+						enum FORWARD;
+						enum DROP;
+						enum REDIRECT;
+					}
+				}
+
+				leaf MIRROR_ACTION {
+					type string {
+						pattern '[a-zA-Z0-9]{1}([-a-zA-Z0-9_]{1,63})' {
+							error-app-tag mirror-action-invalid;
+						}
+					}
+				}
+
+				leaf IP_TYPE {
+					mandatory true;
+					type enumeration {
+						enum ANY;
+						enum IP;
+						enum IPV4;
+						enum IPV4ANY;
+						enum NON_IPV4;
+						enum IPV6ANY;
+						enum NON_IPV6;
+					}
+				}
+
+				leaf IP_PROTOCOL {
+					type uint8 {
+						range "1|2|6|17|46|47|51|103|115";
+					}
+				}
+
+				leaf ETHER_TYPE {
+					type string {
+						pattern "(0x88CC)|(0x8100)|(0x8915)|(0x0806)|(0x0800)|(0x86DD)|(0x8847)" {
+							error-message "Invalid ACL Rule Ether Type";
+							error-app-tag ether-type-invalid;
+						}
+					}
+				}
+
+				choice ip_src_dst {
+					case ipv4_src_dst {
+						when "boolean(IP_TYPE[.='ANY' or .='IP' or .='IPV4' or .='IPV4ANY'])";
+						leaf SRC_IP {
+							mandatory true;
+							type inet:ipv4-prefix;
+						}
+						leaf DST_IP {
+							mandatory true;
+							type inet:ipv4-prefix;
+						}
+					}
+					case ipv6_src_dst {
+						when "boolean(IP_TYPE[.='ANY' or .='IP' or .='IPV6' or .='IPV6ANY'])";
+						leaf SRC_IPV6 {
+							mandatory true;
+							type inet:ipv6-prefix;    
+						}
+						leaf DST_IPV6 {
+							mandatory true;
+							type inet:ipv6-prefix;    
+						}
+					}
+				}
+
+				choice src_port {
+					case l4_src_port {
+						leaf L4_SRC_PORT {
+							type uint16;
+						}
+					}
+					case l4_src_port_range {
+						leaf L4_SRC_PORT_RANGE {
+							type string {
+								pattern "[0-9]{1,5}(-)[0-9]{1,5}" {
+									error-app-tag src-port-range-invalid;
+								}
+							}
+						}
+					}
+				}
+
+				choice dst_port {
+					case l4_dst_port {
+						leaf L4_DST_PORT {
+							type uint16;
+						}
+					}
+					case l4_dst_port_range {
+						leaf L4_DST_PORT_RANGE {
+							type string {
+								pattern "[0-9]{1,5}(-)[0-9]{1,5}" {
+									error-app-tag dst-port-range-invalid;
+								}
+							}
+						}
+					}
+				}
+
+				leaf TCP_FLAGS { 
+					type string {
+						pattern "0[xX][0-9a-fA-F]{2}[/]0[xX][0-9a-fA-F]{2}" {
+							error-app-tag tcp-flag-invalid;
+						}
+					}
+				}
+
+				leaf DSCP { 
+					type uint8 {
```

Comment:
I would like to avoid taking typedef from standard YANG models like IETF/OC. It is better to define our own typedef if needed so that we have the control for custom SONiC YANG.

Reply:
This is not an enum. Can you please explicitly state, why we should not use standard definition here?

Reply:
Let say for example OC supports DSCP range of 0 to 63 and SONiC supports 0..52. How will we handle that ?

Reply:
Prefer we use standard still and we can add constraint on top as needed,

This is what we stated in our guideline and we wrote the guidelines (and took a while to do it too :-) )

Reply:
Once we define dscp from derived type inet:dscp, we can't define error-app-tag for this. We will have to rely on CVL error code. Hope this is fine.

Reply:
**Fixed**